### ISSUE TITLE
protoparse: desc.Descriptor.GetSourceCodeInfo() always returns nil

### ISSUE THREAD

**Issue by limpo1989** [jhump/protoreflect#132]
I tried to modify the code to make it work：
```golang
       // parser.go:147
       // Add this Code to Fix.
	if p.IncludeSourceCodeInfo {
		for _, pr := range protos {
			pr.fd.SourceCodeInfo = pr.generateSourceCodeInfo()
		}
	}
	linkedProtos, err := newLinker(protos).linkFiles()
	if err != nil {
		return nil, err
	}
	fds := make([]*desc.FileDescriptor, len(filenames))
	for i, name := range filenames {
		fd := linkedProtos[name]
		//if p.IncludeSourceCodeInfo {
		//	pr := protos[name]
		//	fd.AsFileDescriptorProto().SourceCodeInfo = pr.generateSourceCodeInfo()
		//}
		fds[i] = fd
	}
```

**Comment by limpo1989** [comment]
#132 this issue not be fix.
Change parse.go to fix it.
```golang
	if p.IncludeSourceCodeInfo {
		for _, pr := range protos {
			pr.fd.SourceCodeInfo = pr.generateSourceCodeInfo()
		}
	}

        linkedProtos, err := newLinker(protos).linkFiles()
```

**Comment by jhump** [comment]
@limpo1989, sorry I closed this prematurely. I didn't quite understand the issue. I had tests to verify something, so I thought everything was working. Looks #160 is a duplicate of this.

**Comment by jhump** [comment]
BTW, the suggested fix dos not work because the logic to interpret options relies on "rich" linked descriptors (to make it feasible to traverse dependencies and resolve custom option names). So if we compute source info first, we'll have source info attributed to uninterpreted options which do not exist in the final descriptor (since they are replaced with actual option values during linking, which has a step that interprets options). Ironically, this would cause it to mirror protoc's old behavior (which used to have the same problem before I submitted a fix: protocolbuffers/protobuf#3819).

I have ideas on a proper fix and can hopefully have a PR soon.
